cmd/sonarExecuteScan.go

```diff
@@ -0,0 +1,242 @@
+package cmd
+
+import (
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/SAP/jenkins-library/pkg/command"
+	piperhttp "github.com/SAP/jenkins-library/pkg/http"
+	"github.com/SAP/jenkins-library/pkg/log"
+	FileUtils "github.com/SAP/jenkins-library/pkg/piperutils"
+	SliceUtils "github.com/SAP/jenkins-library/pkg/piperutils"
+	"github.com/SAP/jenkins-library/pkg/telemetry"
+)
+
+type sonarSettings struct {
+	Binary      string
+	Environment []string
+	Options     []string
+}
+
+var sonar sonarSettings
+
+var execLookPath = exec.LookPath
+var fileUtilsExists = FileUtils.FileExists
+var fileUtilsUnzip = FileUtils.Unzip
+var osRename = os.Rename
+var osRemove = os.Remove
+
+func sonarExecuteScan(options sonarExecuteScanOptions, telemetryData *telemetry.CustomData) error {
+	c := command.Command{}
+	// reroute command output to loging framework
+	c.Stdout(log.Entry().Writer())
+	c.Stderr(log.Entry().Writer())
+
+	client := piperhttp.Client{}
+	client.SetOptions(piperhttp.ClientOptions{Timeout: time.Second * 180})
+
+	sonar = sonarSettings{
+		Binary:      "sonar-scanner",
+		Environment: []string{},
+		Options:     []string{},
+	}
+
+	runSonar(options, &c, &client)
+	return nil
+}
+
+func runSonar(options sonarExecuteScanOptions, runner execRunner, client piperhttp.Downloader) {
+	// Provided by withSonarQubeEnv:
+	// - SONAR_CONFIG_NAME
+	// - SONAR_EXTRA_PROPS
+	// - SONAR_HOST_URL
+	// - SONAR_AUTH_TOKEN
+	// - SONARQUBE_SCANNER_PARAMS = { "sonar.host.url" : "https:\/\/sonar", "sonar.login" : "******"}
+	// - SONAR_MAVEN_GOAL
+	if len(options.Host) > 0 {
+		sonar.Environment = append(sonar.Environment, "SONAR_HOST_URL="+options.Host)
+	}
+	//TODO: SONAR_AUTH_TOKEN or SONAR_TOKEN, both seem to work
+	if len(options.Token) > 0 {
+		sonar.Environment = append(sonar.Environment, "SONAR_TOKEN="+options.Token)
+	}
+	if len(options.Organization) > 0 {
+		sonar.Options = append(sonar.Options, "sonar.organization="+options.Organization)
+	}
+	if len(options.ProjectVersion) > 0 {
+		sonar.Options = append(sonar.Options, "sonar.projectVersion="+options.ProjectVersion)
+	}
+
+	handlePullRequest(options)
+
+	loadSonarScanner(options.SonarScannerDownloadURL, client)
+
+	loadCertificates(runner, "", client)
+
+	log.Entry().
+		WithField("command", sonar.Binary).
+		WithField("options", sonar.Options).
+		WithField("environment", sonar.Environment).
+		Debug("Executing sonar scan command")
+
+	runner.SetEnv(sonar.Environment)
+
+	sonar.Options = SliceUtils.Prefix(sonar.Options, "-D")
+	if err := runner.RunExecutable(sonar.Binary, sonar.Options...); err != nil {
```

Comment:
minor: you may return an error here instead of `log.Fatal`, then also the error situation can be tested more easily.
Fatal handling would then be in your `sonarExecuteScan` function

Reply:
done